### ISSUE TITLE
Why is the BaseType of Power decimal?

### ISSUE THREAD

**Issue by eriove** [anjdreas/UnitsNet#163]
Just noted that `Power` has the `BaseType` `decimal`?  while all other units except `Information` seems to have `double` as their `BaseType`.

Is there a particular reason for this? Since both `Energry` and `Duration` have the `BaseType` `double` is seems reasonable that `Power` also has this `BaseType`.

I noticed this by trying to pass `double.MaxValue` into `Power.FromWatts()`. Perhaps it is a good idea to add a `.MaxValue` and a `.MinValue` to all units of measure? This would also mean that code won't break if the underlying base unit changes.

**Comment by anjdreas** [comment]
**tl;dr**
Yes, Max/MinValue is probably a good idea, if you add it I'll merge it in.
Also I'm open to better ways to solve the whole conundrum of performance vs accurate conversions.

The reason was basically that the generated tests failed the precision requirement of less than `1e-5` error after roundtrip from any unit to base unit and back.

Although that error margin was arbitrarily chosen as the tightest that almost all units fit inside, I had a couple of unit classes where `double` would simply not do, due to the span of its units, such as `Information` units `Exabit` vs `Bit`, and apparently `Power` (I had forgot it was `decimal`).

For `Information` I chose `long`, since `Bit` would never be represented in fractions of bits.
For a couple of other units, `decimal` was the fallback.

I also added the option of overriding test tolerance for single units, such as `protected virtual double CentimeterPerSecondSquaredTolerance { get { return 1e-5; } }` in `AccelerationTestsBase`, but this does not seem to be used currently. I think I started with that approach, then wound up using different underlying types instead to avoid sacrificing conversion accuracy.

Adding Max/MinValue is probably a good idea, if we are to stick with different underlying types. I do question, however, if there are better solutions to this. It feels a bit dirty. We could simply used `decimal` on all units, but I believe the performance is drastically worse than with `double`, although I don't have any numbers on this. Another option is to let the user decide, perhaps as a separate nuget with all units as decimal and the original nuget as double, or even in the same nuget with names like `LengthD` for `Length` with decimal underlying type, but.. I don't feel particularly clean wiping myself with that towel.

On the other end, there are few issues with the current approach - it works, so there needs to be a good reason to change it.

**Comment by eriove** [comment]
I've made a PR with the Min and Max values (#164).

Femtowatt is a quite small number.

If we didn't add the overloads performance wouldn't be an issue, since we are using Power in performance critical code it _could_ be an issue. Once we have all our code hooked up I will run some test with the base type changed to double to see if it actually affect performance.

[MathNet.Numerics](https://github.com/mathnet/mathnet-numerics) uses different namespaces for the concrete implementations of `Matrix<double>` and `Matrix<float>`. That works quite well. But it is easier when it is classes and not structs since generic base types/interface works without boxing.

In this case it wouldn't be as clean. To keep the library size down two different Nuget packages would be needed and that would make it harder to use the library for the average user. I would suggest that we check what the performance measurements say before doing anything at all.

**Comment by anjdreas** [comment]
Where are we with this?

**Comment by anjdreas** [comment]
Closing, will reopen if there are any further discussions.
